v2/cn/blog-jd-page-service-based-on-resty.md

```diff
@@ -0,0 +1,199 @@
+
+1、数据异构，如判断库存状态依赖的套装、配件关系我们进行了异构，未来可以对商家运费等数据进行异构，减少接口依赖；
+
+2、服务闭环，所有单品页上用到的核心接口都接入统一服务；有些是查库/缓存然后做一些业务逻辑，有些是http接口调用然后进行简单的数据逻辑处理；还有一些就是做了下简单的代理，并监控接口服务质量。
+
+## 引入Nginx接入层
+
+我们在设计系统时需要把一些逻辑尽可能前置以此来减轻后端核心逻辑的压力，另外如服务升级/服务降级能非常方便的进行切换，在接入层我们做了如下事情：
+
+* 数据校验/过滤逻辑前置、缓存前置、业务逻辑前置
+* 降级开关前置
+* AB测试
+* 灰度发布/流量切换
+* 监控服务质量
+* 限流
+
+### 数据校验/过滤逻辑前置
+
+我们服务有两种类型的接口：一种是用户无关的接口，另一种则是用户相关的接口；因此我们使用了两种类型的域名c.3.cn/c0.3.cn/c1.3.cn和cd.jd.com；当我们请求cd.jd.com会带着用户cookie信息到服务端；在我们服务器上会进行请求头的处理，用户无关的所有数据通过参数传递，在接入层会丢弃所有的请求头(保留gzip相关的头)；而用户相关的会从cookie中解出用户信息然后通过参数传递到后端；也就是后端应用从来就不关心请求头及Cookie信息，所有信息通过参数传递。
+请求进入接入层后，会对参数进行校验，如果参数校验不合法直接拒绝这次请求；我们对每个请求的参数进行了最严格的数据校验处理，保证数据的有效性。如下所示，我们对关键参数进行了过滤，如果这些参数不合法就直接拒绝请求。
+
+<html>
+<img src="/images/blog/jd-page-service-05.png" width="500">
+</html>
+
+另外我们还会对请求的参数进行过滤然后重新按照固定的模式重新拼装URL调度到后端应用，此时URL上的参数是固定的而且是有序的，可以按照URL进行缓存。
+
+### 缓存前置
+
+我们把很多缓存前置到了接入层，来进行热点数据的削峰，而且配合一致性哈希可能提升缓存的命中率。在缓存时我们按照业务来设置缓存池，减少相互之间的影响和提升并发。我们使用Lua读取共享字典来实现本地缓存。
+
+### 业务逻辑前置
+
+我们在接入层直接实现了一些业务逻辑，原因是当在高峰时出问题，可以在这一层做一些逻辑升级；我们后端是Java应用，当修复逻辑时需要上线，而一次上线可能花费数十秒时间启动应用，重启应用后Java应用JIT的问题会存在性能抖动的问题，可能因为重启造成服务一直启动不起来的问题；而在Nginx中做这件事情，改完代码推送到服务器，重启只需要秒级，而且不存在抖动的问题。这些逻辑都是在Lua中完成。
+
+### 降级开关前置
+
+我们降级开关分为这么几种：接入层开关和后端应用开关、总开关和原子开关；我们在接入层设置开关的目的是防止降级后流量还无谓的打到后端应用；总开关是对整个服务降级，比如库存服务默认有货；而原子开关时整个服务中的其中一个小服务降级，比如库存服务中需要调用商家运费服务，如果只是商家运费服务出问题了，此时可以只降级商家运费服务。另外我们还可以根据服务重要程度来使用超时自动降级机制。
+我们使用init_by_lua_file初始化开关数据，共享字典存储开关数据，提供API进行开关切换（switch_get(“stock.api.not.call”) ~= “1”）。可以实现：秒级切换开关、增量式切换开关（可以按照机器组开启，而不是所有都开启）、功能切换开关、细粒度服务降级开关、非核心服务可以超时自动降级。
+
+比如双十一期间我们有些服务出问题了，我们进行过大服务和小服务的降级操作，这些操作对用户来说都是无感知的。
+
+### AB测试
+
+对于服务升级，最重要的就是能做AB测试，然后根据AB测试的结果来看是否切新服务；而有了接入层非常容易进行这种AB测试；不管是上线还是切换都非常容易。可以在Lua中根据请求的信息调用不同的服务或者upstream分组即可完成AB测试。
+
+### 灰度发布/流量切换
+
+对于一个灵活的系统来说，能随时进行灰度发布和流量切换是非常重要的一件事情，比如验证新服务器是否稳定，或者验证新的架构是否比老架构更优秀，有时候只有在线上跑着才能看出是否有问题；我们在接入层可以通过配置或者写Lua代码来完成这件事情，灵活性非常好。可以设置多个upstream分组，然后根据需要切换分组即可。
+
+### 监控服务质量
+
+对于一个系统最重要的是要有双眼睛能盯着系统来尽可能早的发现问题，我们在接入层会对请求进行代理，记录status、request_time、response_time来监控服务质量，比如根据调用量、状态码是否是200、响应时间来告警。
+
+### 限流
+
+我们系统中存在的主要限流逻辑是：对于大多数请求按照IP请求数限流，对于登陆用户按照用户限流；对于读取缓存的请求不进行限流，只对打到后端系统的请求进行限流。还可以限制用户访问频率，比如使用ngx_lua中的ngx.sleep对请求进行休眠处理，让刷接口的速度降下来；或者种植cookie token之类的，必须按照流程访问。当然还可以对爬虫/刷数据的请求返回假数据来减少影响。
+
+## 前端业务逻辑后置
+
+前端JS应该尽可能少的业务逻辑和一些切换逻辑，因为前端JS一般推送到CDN，假设逻辑出问题了，需要更新代码上线，推送到CDN然后失效各个边缘CDN节点；或者通过版本号机制在服务端模板中修改版本号上线，这两种方式都存在效率问题，假设处理一个紧急故障用这种方式处理完了可能故障也恢复了。因此我们的观点是前端JS只拿数据展示，所有或大部分逻辑交给后端去完成，即静态资源CSS/JS CDN，动态资源JSONP；前端JS瘦身，业务逻辑后置。
+在双十一期间我们的某些服务出问题了，不能更新商品信息，此时秒杀商品需要打标处理，因此我们在服务端完成了这件事情，整个处理过程只需要几十秒就能搞定，避免了商品不能被秒杀的问题。而如果在JS中完成需要耗费非常长的时间，因为JS在客户端还有缓存时间，而且一般缓存时间非常长。
+
+## 前端接口服务端聚合
+
+商品详情页上依赖的服务众多，一个类似的服务需要请求多个不相关的服务接口，造成前端代码臃肿，判断逻辑众多；而我无法忍受这种现状，我想要的结果就是前端异步请求我的一个API，我把相关数据准备好发过去，前端直接拿到数据展示即可；所有或大部分逻辑在服务端完成而不是在客户端完成；因此我们在接入层使用Lua协程机制并发调用多个相关服务然后最后把这些服务进行了合并。
+
+比如推荐服务：最佳组合、推荐配件、优惠套装；通过http://c.3.cn/recommend?methods=accessories,suit,combination&sku=1159330&cat=6728,6740,12408&lid=1&lim=6进行请求获取聚合的数据，这样原来前端需要调用三次的接口只需要一次就能吐出所有数据。
```

Comment:
IF add spaces around URI, when the file translated to html page with pandoc,  encounter one problem: the URI not a string any more but a link that not expected. 
Is it OK or how to fix this?  Thanks.